### ISSUE TITLE
Logging function not replacing format placeholder

### ISSUE THREAD

**Issue by AnHeuermann** [ThummeTo/FMI.jl#69]
## Description

The log function provided by FMI.jl is not replacing format placeholders with the actual string.

## How to reproduce

When initializing a Pendulum FMU and enabling `loggingOn = true`:

```julia
julia> using FMI
julia> fmu = fmiLoad("Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum.fmu");
julia> fmiInstantiate!(fmu; loggingOn = true);
[ Info: [OK][logAll][Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum]: fmi2Instantiate: Trying to find simulation settings %s.
[ Info: [OK][logAll][Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum]: fmi2Instantiate: Using default simulation settings.
[ Info: [OK][logFmi2Call][Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum]: fmi2Instantiate: GUID=%s
```

I would expect `%s` to be replaced by the GUID string.

FMU (remove the .zip from name): 
[Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum.fmu.zip](https://github.com/ThummeTo/FMI.jl/files/7855658/Modelica.Mechanics.MultiBody.Examples.Elementary.Pendulum.fmu.zip)
 

## Version 'n stuff

Generation Tool: OpenModelica v1.19.0-dev.505+ga499198eb3
OS: Ubuntu 20.04
Julia:
```
Julia Version 1.7.1
Commit ac5cc99908 (2021-12-22 19:35 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-10700KF CPU @ 3.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
```
```
(@v1.7) pkg> status FMI
      Status `~/.julia/environments/v1.7/Project.toml`
  [14a09403] FMI v0.3.6
```

**Comment by ThummeTo** [comment]
This is currently not possible, because the Julia-C-Pipeline currently does not support variadic arguements. The only way to currently solve this, would be to redirect the variadic C function call to a DLL/LIB, replace the format placeholders inside and call a Julia-Function with the formatted string (no variadic arguments any more). As said, this would need us to ship a DLL/LIB  with the library or build a DLL/LIB on the system that implements this single function.

Funny thing: In some eraly version we already tried this and it works fine, but the preferable solution would be to handle the variadic callback function with "native" Julia.

PS: The only tool I currently know that uses the variadic argument calling convention is OpenModelica, Dymoal and many others just deliver a ready to use formatted string.

So the options would be to deliver some C-library-code and (automated) compile it for the target machine and redirect the callback function or wait for Julia to support variadic arguments in c-functions.

**Comment by AnHeuermann** [comment]
If other tools would struggle with this I would just change it in OpenModelica. But the tools I tested (OMSimulator, FMPy) can handle it. Not sure how python does FMPy does this. Maybe Python can do this or they added a similar binary.

> So the options would be to deliver some C-library-code and (automated) compile it for the target machine and redirect the callback function (which is not a really clean solution) or wait for Julia to support variadic arguments in c-functions.

The C-library would be the fastest option. Do you think you can add that library to FMI.jl until there is a pure Julia solution available?

**Comment by AnHeuermann** [comment]
Sure, I can check it out on Linux.

I think the recommended way for C dependencies is to use [BinaryBuilder.jl](https://github.com/JuliaPackaging/BinaryBuilder.jl) and [Yggrasil](https://github.com/JuliaPackaging/Yggdrasil). I never tried them myself, but instead added pre-build binaries to a Julia package. But then you have the trouble to compile the sources for all available systems.

**Comment by ThummeTo** [comment]
The more I think about it, the more I don't like the idea of checking in such a "hacky" solution. Further, we would build platform-dependent code around a library that is platform-independent, which seems not a good idea.

Because - if I understand you right - this issue isn't really urgent (there is the option to export the FMU with ready-formatted strings), I would suggest to wait for a language native fix (variadic c-functions), instead of deploying a (propably) unreliable workaround...

What do you think?

**Comment by AnHeuermann** [comment]
> The more I think about it, the more I don't like the idea of checking in such a "hacky" solution. 

I'm not sure it is a "hacky" solution. Not everything works in Julia and you are dealing with C FMUs, so havon some C sources / binaries for FMI.jl sounds fine.

> this issue isn't really urgent (there is the option to export the FMU with ready-formatted strings)
I don't know about an option for OpenModelica to have ready-formatted strings. I could add it, but the FMUs are according to the specification, so I don't think it makes that much sense to change OM FMUs.
But yes, not an urgent problem.

But how long would a language native fix be? Do you know any discussion about this? I couldn't find much about giving variadic Julia functions to a C function. Only the other way around.

**Comment by ThummeTo** [comment]
I just found an article, that this should be possible using CBindings.jl. 
I will check this out ...

**Comment by ThummeTo** [comment]
This is fixed for Windows in FMI.jl 0.6.0. Until Julia supports variadic c-calls, you can use `fmiInstantiate(....; externalCallbacks=true)`.

**Comment by ThummeTo** [comment]
To solve this the way it should be solved, it is needed to offer a variadic function for `fmi2CallbackLogger`, that formats the message like a call to `sprintf` (or `printf`). In native Julia, variadic c-functions are not supported by now (as far as I know). A possible option my be using [CBindings.jl](https://github.com/analytech-solutions/CBinding.jl) but I am not familiar with ...

**Comment by AnHeuermann** [comment]
I added a binaries for Linux and Mac in https://github.com/ThummeTo/FMIImport.jl/pull/40.

**Comment by ThummeTo** [comment]
Much appreciated! I will merge them soon together with some other changes.
Thanks!

**Comment by ThummeTo** [comment]
Closed by FMIImport v0.11.0 (FMI.jl 0.10.0) via [PR#45](https://github.com/ThummeTo/FMIImport.jl/pull/45)
